### PR TITLE
Override J9::Compilation::needRelocationsForStatics

### LINKED ISSUE #6688: compilationAotStaticFieldReloFailure seen with JITServer
Daytrader7 experiments with JITServer showed quite a few compilation failures with an error code of `compilationAotStaticFieldReloFailure`. These experiments do not use any AOT.

E.g.:
```
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=714043us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=731 MB mem=[region=320 system=16384]KB
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=533537us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=731 MB mem=[region=192 system=16384]KB
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=477081us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=730 MB mem=[region=192 system=16384]KB
! com/ibm/crypto/provider/bf.newInstance(Ljava/lang/Object;)Ljava/lang/Object; time=57875us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=729 MB mem=[region=128 system=16384]KB
! com/ibm/crypto/provider/bf.newInstance(Ljava/lang/Object;)Ljava/lang/Object; time=57524us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=729 MB mem=[region=128 system=16384]KB
! com/ibm/crypto/provider/bf.newInstance(Ljava/lang/Object;)Ljava/lang/Object; time=86312us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=729 MB mem=[region=128 system=16384]KB
! com/ibm/crypto/provider/bf.newInstance(Ljava/lang/Object;)Ljava/lang/Object; time=90244us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=729 MB mem=[region=128 system=16384]KB
! javax/crypto/Mac.a(Ljava/security/Key;Ljava/security/spec/AlgorithmParameterSpec;)V time=142767us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=725 MB mem=[region=128 system=16384]KB
! javax/crypto/Mac.a(Ljava/security/Key;Ljava/security/spec/AlgorithmParameterSpec;)V time=151178us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=725 MB mem=[region=128 system=16384]KB
! javax/crypto/Mac.a(Ljava/security/Key;Ljava/security/spec/AlgorithmParameterSpec;)V time=174353us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=724 MB mem=[region=128 system=16384]KB
! javax/crypto/Mac.a(Ljava/security/Key;Ljava/security/spec/AlgorithmParameterSpec;)V time=130471us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=724 MB mem=[region=128 system=16384]KB
! javax/json/spi/JsonProvider.provider()Ljavax/json/spi/JsonProvider; time=82663us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=724 MB mem=[region=128 system=16384]KB
! javax/json/spi/JsonProvider.provider()Ljavax/json/spi/JsonProvider; time=55908us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=724 MB mem=[region=128 system=16384]KB
! javax/json/spi/JsonProvider.provider()Ljavax/json/spi/JsonProvider; time=89522us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=724 MB mem=[region=128 system=16384]KB
! javax/json/spi/JsonProvider.provider()Ljavax/json/spi/JsonProvider; time=56711us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=724 MB mem=[region=128 system=16384]KB
! sun/security/jca/GetInstance.getInstance(Ljava/security/Provider$Service;Ljava/lang/Class;)Lsun/security/jca/GetInstance$Instance; time=88233us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=721 MB mem=[region=128 system=16384]KB
! sun/security/jca/GetInstance.getInstance(Ljava/security/Provider$Service;Ljava/lang/Class;)Lsun/security/jca/GetInstance$Instance; time=84928us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=721 MB mem=[region=128 system=16384]KB
! sun/security/jca/GetInstance.getInstance(Ljava/security/Provider$Service;Ljava/lang/Class;)Lsun/security/jca/GetInstance$Instance; time=90771us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=721 MB mem=[region=128 system=16384]KB
! sun/security/jca/GetInstance.getInstance(Ljava/security/Provider$Service;Ljava/lang/Class;)Lsun/security/jca/GetInstance$Instance; time=61146us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=721 MB mem=[region=128 system=16384]KB
! com/ibm/ws/sib/msgstore/task/TaskList.commit(Lcom/ibm/ws/sib/msgstore/transactions/impl/PersistentTransaction;)V time=63114us compilationInterrupted memLimit=262144 KB freePhysicalMemory=710 MB mem=[region=64 system=16384]KB
! com/ibm/ws/sib/utils/ras/SibTr.entry(Lcom/ibm/websphere/ras/TraceComponent;Ljava/lang/String;Ljava/lang/Object;)V time=22070us compilationInterrupted memLimit=262144 KB freePhysicalMemory=710 MB mem=[region=128 system=16384]KB
! com/ibm/ws/session/SessionEventDispatcher.sessionLiveCountInc(Ljava/lang/Object;)V time=20301us compilationInterrupted memLimit=262144 KB freePhysicalMemory=710 MB mem=[region=64 system=16384]KB
! com/ibm/ws/sib/mfp/impl/JsDestinationAddressImpl.getDestinationName()Ljava/lang/String; time=38727us compilationInterrupted memLimit=262144 KB freePhysicalMemory=710 MB mem=[region=64 system=16384]KB
! com/ibm/ws/sib/msgstore/task/TaskList.postComplete(Lcom/ibm/ws/sib/msgstore/transactions/impl/PersistentTransaction;Z)V time=55662us compilationInterrupted memLimit=262144 KB freePhysicalMemory=710 MB mem=[region=64 system=16384]KB
! com/ibm/ws/sib/msgstore/persistence/objectManager/UniqueKeyGeneratorImpl.getUniqueValue()J time=14548us compilationInterrupted memLimit=262144 KB freePhysicalMemory=710 MB mem=[region=64 system=16384]KB
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=423015us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=603 MB mem=[region=192 system=16384]KB
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=422623us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=603 MB mem=[region=192 system=16384]KB
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=439139us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=602 MB mem=[region=192 system=16384]KB
! javax/crypto/Cipher.getInstance(Ljava/lang/String;)Ljavax/crypto/Cipher; time=580999us compilationAotStaticFieldReloFailure memLimit=262144 KB freePhysicalMemory=603 MB mem=[region=192 system=16384]KB
```

### PR DESCRIPTION
**Pull request by dmitry-ten**
A while ago I added a new query `needRelocationsForStatics`
to OMR, which always returns true.
The plan was to override that query in OpenJ9 so that we don't
require relocation of statics for regular JITServer compilations.
For some reason I did not do that, which led to issue #6688,
i.e. we add a relocation for unresolved static field during a
non-AOT compilation, fail to relocate (because unresolved), and
fail compilation. Fixing this now.
Also, remove a similar query `J9::CodeGenerator::needRelocationsForStatics`, since it does the same thing but comp object is better suited for it.

Fixes #6688

### CODE REVIEW

**Comment by dmitry-ten** [comment]
@mpirvu could you review?
For reference, [this](https://github.com/eclipse/omr/pull/2878) is the PR I'm talking about.

**Comment by dmitry-ten** [comment]
[Tests](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/3040/)

**Comment by dmitry-ten** [comment]
Yeah, I agree, it's already confusing. Shouldn't have added that new query in the first place.
Then I think we should close this PR and make another PR into OMR to change code there.

**Comment by mpirvu** [comment]
I think that the right approach is to modify `compiler/x/codegen/X86BinaryEncoding.cpp` to test `cg->needRelocationsForStatics()` instead of `comp->needRelocationsForStatics()` and remove the definition of `needRelocationsForStatics()` from the `OMRCompilation.hpp` altogether.
All these changes are contained in omr.
@mstoodle I would like your opinion as well please, given your involvement with the original PR: https://github.com/eclipse/omr/pull/2878

**Comment by dmitry-ten** [comment]
Also, looking at the code for `x/codegen/X86BinaryEncoding.cpp` I wonder if we could remove some of the `#ifdef J9_PROJECT_SPECIFIC` using the same approach as suggested by Marius.
For example we could replace `comp->fej9()->needClassAndMethodPointerRelocations()` with the same query from OMR code generator. Assuming the access to front-end is the only thing for which `ifdef` is required.

**Comment by mstoodle** [comment]
For `comp` versus `cg`:  is a particular platform ever likely to give a different answer to that query? If there isn't anything platform-specific about the answer, I think `comp` is a better home for the query. If the answer is platform-specific (which doesn't seem to be the case) then `cg` would be a better place for it.

**Comment by dmitry-ten** [comment]
I don't think it's platform-specific. @dsouzai maybe you would know better?
If `comp` object is a better place, then other similar queries in `cg` should be moved there as well.

**Comment by dsouzai** [comment]
It's before my time, but I believe the motivation for that query was to remove all of the `isAOT` queries and replace them a query that better reflected the context in which it was being used. I think whether it's in the `comp` or `cg` depends on the context, for example

```
bool OMR::CodeGenerator::needGuardSitesEvenWhenGuardRemoved() { return self()->comp()->compileRelocatableCode(); }
```

However, what @mstoodle says also makes sense, in that it could also be used for when different platforms might return different answers. However, as the code stands, all "isAOT" queries return the same thing on all platform (namely, `true` if an AOT compile, and `false` otherwise) regardless of platform.

So tl;dr, put the query wherever it seems best suited I guess.

**Comment by dmitry-ten** [comment]
I don't have any preference either way

**Comment by mpirvu** [comment]
I am ok with comp instead of cg as well.
This means that this PR could stay as is and another PR be opened to replace all `cg->needRelocationsForStatics()` with `cg->needRelocationsForStatics()`

**Comment by dmitry-ten** [comment]
Removed `J9::CodeGenerator::needRelocationsForStatics` and made an OMR PR: eclipse/omr#4197

**Comment by dmitry-ten** [comment]
I already removed all usages from OMR. 
Do we have to do in 3 PRs? Can't we just merge at the same time?

**Comment by mpirvu** [comment]
If you do (2) before (1), then openj9 code is going to use the default implementation provided by OMR which returns `false` even for AOT. So AOT in master branch is going to be broken for some window of time until (1) is delivered. Synchronizing these changes perfectly across repos is not possible.

**Comment by dmitry-ten** [comment]
Ok, I'll do it then. Closing this now to reopen in master.

**Comment by dchopra001** [comment]
@dmitry-ten This query will probably need to return `true` on Z since we may use a PC-relative load for the static.

**Comment by dmitry-ten** [comment]
Then it actually makes sense to keep the query in codegen...
Unless someone has another idea.

**Comment by mpirvu** [comment]
> Then it actually makes sense to keep the query in codegen...

Right, we should go back to the original plan
